Amazon.QLDB.Driver/result/AsyncResult.cs

```diff
@@ -16,23 +16,141 @@ namespace Amazon.QLDB.Driver
+    ///
+    /// Note that due to the fact that a result can only be retrieved from QLDB once, the Result may only be iterated
+    /// over once. Attempts to do so multiple times will result in an exception.
+    ///
+    /// This implementation should be used by default to avoid excess memory consumption and to improve performance.
+    /// </summary>
     internal class AsyncResult : IAsyncResult
     {
+        private readonly IonAsyncEnumerator ionEnumerator;
+        private bool isRetrieved = false;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AsyncResult"/> class.
+        /// </summary>
+        ///
+        /// <param name="session">The parent session that represents the communication channel to QLDB.</param>
+        /// <param name="statementResult">The result of the statement execution.</param>
+        /// <param name="txnId">The unique ID of the transaction.</param>
+        internal AsyncResult(Session session, string txnId, ExecuteStatementResult statementResult)
+        {
+            this.ionEnumerator = new IonAsyncEnumerator(session, txnId, statementResult);
+        }
+
+        /// <inheritdoc/>
         public IAsyncEnumerator<IIonValue> GetAsyncEnumerator(CancellationToken cancellationToken = default)
         {
-            throw new NotImplementedException();
+            if (this.isRetrieved)
+            {
+                throw new InvalidOperationException();
+            }
+
+            this.isRetrieved = true;
+            return this.ionEnumerator;
         }
 
+        /// <summary>
+        /// Gets the current query statistics for the number of read IO requests. The statistics are stateful.
+        /// </summary>
+        ///
+        /// <returns>The current IOUsage statistics.</returns>
         public IOUsage? GetConsumedIOs()
         {
-            throw new NotImplementedException();
+            return this.ionEnumerator.GetConsumedIOs();
         }
 
+        /// <summary>
+        /// Gets the current query statistics for server-side processing time. The statistics are stateful.
+        /// </summary>
+        ///
+        /// <returns>The current TimingInformation statistics.</returns>
         public TimingInformation? GetTimingInformation()
         {
-            throw new NotImplementedException();
+            return this.ionEnumerator.GetTimingInformation();
+        }
+
+        /// <summary>
+        /// Object which allows for asynchronous iteration over the individual Ion values that make up the whole result of a statement
+        /// execution against QLDB.
+        /// </summary>
+        private class IonAsyncEnumerator : BaseIonEnumerator, IAsyncEnumerator<IIonValue>
+        {
+            private readonly Session session;
+            private readonly string txnId;
+            private new IEnumerator<ValueHolder> currentEnumerator;
+            private string nextPageToken;
```

Comment:
These properties all seem shared with the sync IonEnumerator. I believe they can be moved to the base class.

Reply:
Done.